### DISCUSSION TITLE
EntityManager.create constructor parameter mapping incorrect with defined setter/getter

### DISCUSSION THREAD

**Discussion by DinoCloudlet** [mikro-orm/mikro-orm#4187] [Q&A]
## Entity Definition

```typescript
@Entity()
export class MyEntity {
  @Property( {name: 'myProperty'})
  private _myProperty: number;

  @Property()
  public otherProperty: string;


  constructor(myProperty: number) {
    this.myPropterty = param;
  }

  set myProperty(prop: number) {
    this.validate(prop);
    this._myProperty = prop;
  }

  get myProperty(): number {
    return this._myProperty;
  }
}
```

## EntityManager.create

```typescript
  entityManager.create(MyEntity, {myProperty: 0, otherProperty: 'something else'})
```

# Behaviour

- constructor Parameter myProperty contains an object with all attributes of the create parameter except the myProperty attribute.
```typescript
  {
    otherProperty: 'something else'
  }
```

**Comment by B4nan** [comment]
> `@Property( {name: 'myProperty'})`

This is probably not doing what you think - you are setting the column name to `myProperty` here. The property name is still `_myProperty` and that is what the ORM tries to map to the constructor, but you don't provide such a key in the params. 

But it's weird that it strips the key when it's not matched, will look into that.

**Reply by B4nan** [reply]
Maybe we could allow redefining the property name too, as long as there is corresponding get/set pair defined.

**Reply by DinoCloudlet** [reply]
Sorry for the late response.

Without renaming the column with the decorator parameter "name" queries will throw errors, since mikro orm does not know how to map the column '_myProperty' to 'myProperty' and vice versa. On the other hand the usage of private members leads to the overhead that if you want to protect an attribute from beeing written without validation you have to always define a getter so that the EntityData type can detect the property so it can be used in e.g. the find options of queries.

Thrown Error:
```
Trying to query by not existing property MyEntity.myProperty
```

The main problem behind this is our current usage of mikro orm. We want to put the code for validation and logic that just depends on this single Entity inside the Entity class and protect some of its properties from writing without validation. There would be a huge overhead always writing wrapper classes for each entity or "wrapper commands" for every single validation that is beeing reused.

Currently we are using a workaround by calling the constructor directly and avoid using the assign functionalities of the BaseEntity. In The Documentation could be added that the second EntityManager.create Parameter can be an instance of the Entity. In this case the parameter is not treated as EntityData instead it is used as the Entity instance itself. The constructor will not be called again in this case. (It works the same as calling entitiyManager.persist except that you have control over population and it prevents bugs by forgetting the persist call)

```typescript
entityManager.create(MyEntity, new MyEntity(10));
```
